### PR TITLE
[esl-carousel]: restructuring wave 3

### PR DESCRIPTION
**Pull request by ala-n**

### CODE REVIEW

File: src/modules/esl-carousel/core/esl-carousel.events.ts [hunk 1]
```diff
@@ -36,17 +44,28 @@ export class ESLCarouselSlideEvent extends Event implements ESLCarouselSlideEven
...
 
+/** {@link ESLCarouselChangeEvent} init object */
 interface ESLCarouselChangeEventInit {
-  prop: string;
-  addedSlide?: ESLCarouselSlide;
-  removedSlide?: ESLCarouselSlide;
+  /** Current {@link ESLCarousel} instance config */
+  config: ESLCarouselStaticState;
+  /** Previous {@link ESLCarousel} instance config */
+  oldConfig?: ESLCarouselStaticState;
```
Comment (julia-murashko) [inline src/modules/esl-carousel/core/esl-carousel.events.ts:52]: Mb previousConfig? Seems to fit better

File: src/modules/esl-carousel/core/esl-carousel.renderer.ts [hunk 2]
```diff
@@ -9,20 +9,28 @@ export abstract class ESLCarouselRenderer {
...
-  /**  @returns count of carousel slides. */
+  /** @returns count of carousel slides */
   public get size(): number {
     return this.carousel.size;
   }
+  /** @returns count of visible carousel slides */
+  public get count(): number {
+    return this.carousel.config.count;
+  }
+  /** @returns count of carousel slides */
```
Comment (julia-murashko) [inline src/modules/esl-carousel/core/esl-carousel.renderer.ts:30]:
```suggestion
  /** @returns if the carousel is in a loop mode */
```

File: src/modules/esl-carousel/core/esl-carousel.ts [hunk 3]
```diff
@@ -96,42 +103,45 @@ export class ESLCarousel extends ESLBaseElement implements ESLCarouselState {
...
   }
 
-  /** Updates the config and the state that is associated with. */
-  public update(force: boolean = false): void {
-    const cfgChanged = this.count !== this.countRule.value || this.loop !== this.loopRule.value;
-    const typeChanged = this.renderer?.type !== this.typeRule.value;
+  /** Updates the config and the state that is associated with */
+  @decorate(microtask)
+  public update(): void {
+    const oldType = this.renderer.type;
```
Comment (julia-murashko) [inline src/modules/esl-carousel/core/esl-carousel.ts:109]: minor: mb previousType?

File: src/modules/esl-utils/async/microtask.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/**
+ * A utility decorator to postpone callback execution once after the main task execution
```
Comment (julia-murashko) [inline src/modules/esl-utils/async/microtask.ts:2]:
```suggestion
 * A decorator utility to postpone callback execution once after the main task execution
```

File: src/modules/esl-utils/async/test/microtask.test.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    expect(fn).toBeCalledTimes(0);
+  });
+  test('Decorated as microtask callback called once the macrotask done', async () => {
+    const fn = jest.fn();
+    const decorated = microtask(fn);
+    for (let i = 0; i < 5; i++) decorated();
+    await Promise.resolve();
+    expect(fn).toBeCalledTimes(1);
+  });
+  test('Decorated as microtask callback receive a list of call arguments', async () => {
```
Comment (julia-murashko) [inline src/modules/esl-utils/async/test/microtask.test.ts:17]:
```suggestion
  test('Decorated as microtask callback receives a list of call arguments', async () => {
```

**Review by julia-murashko — APPROVED** [review]

File: src/modules/esl-carousel/core/esl-carousel.events.ts (hunk 1 above)
Comment (ala-n) [reply src/modules/esl-carousel/core/esl-carousel.events.ts:52]: Just think about the similarity to the MutationObserver props, e.g. `oldValue` vs `newValue`
But let's discuss this on tomorrow's call

**Pull request merged**